src/network/networkd-setlink.c

```diff
@@ -689,7 +733,15 @@ int link_request_to_set_mac(Link *link) {
         if (ether_addr_equal(&link->hw_addr.ether, link->network->mac))
                 return 0;
 
-        return link_request_set_link(link, SET_LINK_MAC, link_set_mac_handler, NULL);
+        r = link_request_set_link(link, SET_LINK_MAC,
+                                  allow_retry ? link_set_mac_allow_retry_handler : link_set_mac_handler,
+                                  &req);
+        if (r < 0)
+                return r;
+
+        /* This indicates whether we should bring down the interface or not before setting MAC address. */
+        req->userdata = INT_TO_PTR(!allow_retry);
```

Comment:
hmm, do you actually need to set the userdata here? You could also just compare the handler installed for this `Request` with the pointers to link_set_mac_allow_retry_handler(), right?

i think that would be a bit nicer maybe, since userdata is so opaque